source/request-response/request_response_client.c

```diff
@@ -913,6 +887,139 @@ static void s_apply_publish_to_streaming_operation_list(
     }
 }
 
+static void s_complete_operation_with_correlation_token(
+    struct aws_mqtt_request_response_client *rr_client,
+    struct aws_byte_cursor correlation_token,
+    const struct aws_protocol_adapter_incoming_publish_event *publish_event) {
+    struct aws_hash_element *hash_element = NULL;
+
+    if (aws_hash_table_find(&rr_client->operations_by_correlation_tokens, &correlation_token, &hash_element)) {
+        return;
+    }
+
+    if (hash_element == NULL) {
+        AWS_LOGF_DEBUG(
+            AWS_LS_MQTT_REQUEST_RESPONSE,
+            "id=%p: request-response client incoming publish on response path topic '" PRInSTR
+            "' and correlation token '" PRInSTR "' does not have an originating request entry",
+            (void *)rr_client,
+            AWS_BYTE_CURSOR_PRI(publish_event->topic),
+            AWS_BYTE_CURSOR_PRI(correlation_token));
+        return;
+    }
+
+    struct aws_mqtt_rr_client_operation *operation = hash_element->value;
+    AWS_FATAL_ASSERT(operation->type == AWS_MRROT_REQUEST);
+
+    if (operation->state == AWS_MRROS_PENDING_DESTROY) {
+        AWS_LOGF_DEBUG(
+            AWS_LS_MQTT_REQUEST_RESPONSE,
+            "id=%p: request-response operation %" PRIu64 " cannot be completed, already in pending destruction state",
+            (void *)operation->client_internal_ref,
+            operation->id);
+        return;
+    }
+
+    AWS_LOGF_DEBUG(
```

Comment:
Trivial: Should this log occur if the completion callback is invoked? Is this a failed operation if there's no completion callback and the payload of the correlated publish event isn't delivered anywhere?

Reply:
If it's in pending destroy something else already "successfully" completed it (like a timeout).